### PR TITLE
Try to align the native code to the next 8 byte boundary

### PR DESCRIPTION
**Pull request by filmor**
### What does this implement/fix? Explain your changes.

Aligns the ARM code to the next 8 byte boundary.

### Does this close any currently open issues?

Maybe #898

### CODE REVIEW

File: src/runtime/typemanager.cs [hunk 1]
```diff
@@ -696,8 +696,21 @@ internal static void InitializeNativeCodePage()
...
             IMemoryMapper mapper = CreateMemoryMapper();
             int codeLength = NativeCode.Active.Code.Length;
-            NativeCodePage = mapper.MapWriteable(codeLength);
-            Marshal.Copy(NativeCode.Active.Code, 0, NativeCodePage, codeLength);
+
+            // Request 7 bytes more than required such that we can align to an 8 byte boundary on ARM
+            NativeCodePage = mapper.MapWriteable(codeLength + 7);
+
+            // Round to 8 byte boundary
+            var start = new IntPtr(((NativeCodePage.ToInt64() + 7) >> 3) << 3);
```
Comment (lostmsu) [inline src/runtime/typemanager.cs]: `>> 3) << 3)` can be replaced with `& ~0x7L`
IMHO, even better to write a generic align function (and unit test it)

File: src/runtime/typemanager.cs [hunk 2]
```diff
@@ -698,8 +698,21 @@ internal static void InitializeNativeCodePage()
             // to be executable.
             IMemoryMapper mapper = CreateMemoryMapper();
             int codeLength = NativeCode.Active.Code.Length;
-            NativeCodePage = mapper.MapWriteable(codeLength);
-            Marshal.Copy(NativeCode.Active.Code, 0, NativeCodePage, codeLength);
+
+            // Request 7 bytes more than required such that we can align to an 8 byte boundary on ARM
+            NativeCodePage = mapper.MapWriteable(codeLength + 7);
```
Comment (benoithudson) [inline src/runtime/typemanager.cs]: mmap will return minimum a full page, and pages are typically 4096-byte-aligned (certainly more than 8-byte aligned). Add some logging to make sure, but I don't expect your changes are actually doing anything.

After the bx lr you might need a nop? I don't know this ISA myself, but it's not uncommon in architectures with branch prediction (i.e. basically all of them) that it'll unconditionally run an opcode right after a branch.

**Comment by benoithudson** [comment]
You mentioned compiling on the raspberry pi was slow. You can compile the C# dll on any platform -- as long as you set the `#defines` properly, the generated code is compatible. It's just the platform you run on that has to match your `#defines`. We generally build on windows for all three platforms (windows/mac/linux), I've done it on mac as well.

File: src/runtime/typemanager.cs (hunk 2 above)
Comment (benoithudson) [reply src/runtime/typemanager.cs]: https://community.arm.com/developer/ip-products/processors/f/cortex-m-forum/7154/alignment-in-arm

Looks like you need 16-byte-aligned if it's 64-bit ARM. So your Return1 is not properly aligned.

Comment (filmor) [reply src/runtime/typemanager.cs]: The older RPis are 32bit, though, so this (8 byte alignment) can't be the current issue.

**Pull request closed without merging**